### ISSUE TITLE
The switch expression does not handle all possible inputs - but it does

### ISSUE THREAD

**Issue by MrYossu** [dotnet/roslyn#52081]
**Version Used**: 
Blazor server-side project using .NET5 and C#9

**Steps to Reproduce**:

The following method...

```c#
    public string PartnershipLevel(DateTime date) {
      int maxSharesAtDate = MaxSharesAtDate(date);
      int sumOfSharesAtDate = SumOfSharesAtDate(date);
      return sumOfSharesAtDate == 0
        ? "0%"
        : sumOfSharesAtDate switch {
          int n when n <= maxSharesAtDate => (100 * n / (double)maxSharesAtDate).ToString("F0") + "%",
          int n when n > maxSharesAtDate => ((double)n / maxSharesAtDate).ToString(),
        };
    }
```

**Expected Behavior**:
As the `switch` expression has two clauses, one that handles when `n` is below an integer value and the other than handles when it is equal to or above that value, I would not expect the compiler to issue a warning.

**Actual Behavior**:
I get the following compiler warning...

_Warning CS8846: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '\_' is not covered. However, a pattern with a 'when' clause might successfully match this value._

I understand (from reading the discussion in [this issue](https://github.com/dotnet/roslyn/issues/38571)) that the compiler looks at the underlying type, which is why you can get this warning when using an `Enum`, as that could really be an `int` under the covers, and take a value not covered by the `Enum` itself, but in my case, the `switch` variable is an `int`, and an integer value can only be less than, or greater than or equal to another integer.

Hope that makes sense.

**Comment by Youssef1313** [comment]
> Warning CS8846: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '_' is not covered. However, a pattern with a 'when' clause might successfully match this value.

To me, the message implies that this is by-design. It explicitly says that a pattern with a when clause can match this value.

But I suggest you wait someone from the compilers team to confirm.

**Comment by MrYossu** [comment]
@Youssef1313 Thanks for the reply, but I would have expected the compiler to see that there aren't any `when` clauses in there, and so not give the warning.

**Comment by jaredpar** [comment]
Indeed this is "By Design". The presence of the `when` clause makes it harder for the compiler to determine if the match is exhaustive here. Hence the error message is correct here by the design.

**Comment by MrYossu** [comment]
@jaredpar Thanks for the reply, but my code doesn't have a `when` clause. Surely the compiler can see that and not issue the warning?

**Comment by Youssef1313** [comment]
@MrYossu Your code has a `when` clause (unless you have copied an incorrect code?):

```csharp
sumOfSharesAtDate switch {
          int n when n <= maxSharesAtDate => (100 * n / (double)maxSharesAtDate).ToString("F0") + "%",
          int n when n > maxSharesAtDate => ((double)n / maxSharesAtDate).ToString(),
        };
```

**Comment by MrYossu** [comment]
Duh, so it does! Why did I think it didn't?

OK, fair kop squire, I'll go and do something I'm more suited to, such as gazing out of the window. I think I'd be more productive that way!

Thanks again

**Comment by jaredpar** [comment]
Think this has happened to us all at some point. Glad we were able to figure it out :smile:
